viscy/data/dynacell.py

```diff
@@ -0,0 +1,224 @@
+        self.organelles = organelles
+        self.infection_conditions = infection_conditions
+        self.channel_name = channel_name
+        self.z_slice = z_slice
+        self.database = pd.read_csv(database_path, dtype={"FOV": str})
+        self._process_database()
+
+    def _process_database(self):
+        # Select the portion of the database that matches the provided criteria
+        self._filtered_db = self.database[
+            self.database["Cell type"].isin(self.cell_types)
+            & self.database["Organelle"].isin(self.organelles)
+            & self.database["Infection"].isin(self.infection_conditions)
+        ].copy()
+
+        # Extract zarr store paths
+        self._filtered_db["Zarr path"] = self._filtered_db["Path"].apply(
+            lambda x: Path(*Path(x).parts[:-3])
+        )
+        self._filtered_db = self._filtered_db.drop_duplicates(subset=["Zarr path"])
+
+        # Store values for later use
+        self.zarr_paths = self._filtered_db["Zarr path"].values.tolist()
+        self.cell_types_per_store = self._filtered_db["Cell type"].values.tolist()
+        self.organelles_per_store = self._filtered_db["Organelle"].values.tolist()
+        self.infection_per_store = self._filtered_db["Infection"].values.tolist()
+
+    def __getitem__(self, idx) -> dict:
+        return {
+            "zarr_path": self.zarr_paths[idx],
+            "cell_type": self.cell_types_per_store[idx],
+            "organelle": self.organelles_per_store[idx],
+            "infection_condition": self.infection_per_store[idx],
+            "channel_name": self.channel_name,
+            "z_slice": self.z_slice,
+        }
+
+    def __len__(self) -> int:
+        return len(self.zarr_paths)
+
+
+class DynaCellDataModule(LightningDataModule):
+    def __init__(
+        self,
+        target_database: DynaCellDataBase,
+        pred_database: DynaCellDataBase,
+        batch_size: int,
+        num_workers: int,
+    ) -> None:
+        super().__init__()
+        self.target_database = target_database
+        self.pred_database = pred_database
+        self.batch_size = batch_size
+        self.num_workers = num_workers
+
+    def setup(self, stage: str) -> None:
+        if stage != "test":
+            raise NotImplementedError("Only test stage is supported!")
+
+        # Verify both databases have the same length
+        if len(self.target_database) != len(self.pred_database):
+            raise ValueError(
+                f"Target database length ({len(self.target_database)}) doesn't match "
+                f"prediction database length ({len(self.pred_database)})"
+            )
+
+        # Create datasets
+        datasets = []
+        for i in range(len(self.target_database)):
+            target_data = self.target_database[i]
+            pred_data = self.pred_database[i]
+
+            # Ensure target and prediction metadata match
+            self._validate_matching_metadata(target_data, pred_data, i)
```

Comment:
How long is this loop?

Reply:
for now we don't parallelize. Are you thinking on spitting it as batches?